src/transform/src/map_lifting.rs

```diff
@@ -191,13 +272,47 @@ impl LiteralLifting {
 
                 if scalars.is_empty() {
                     *relation = input.take_dangerous();
+                } else {
+                    // Permute columns to put literals at end, if any, hope project lifted.
+                    let literal_count = scalars.iter().filter(|e| e.is_literal()).count();
```

Comment:
Come to think of it, you can similarly use sort in this branch to automatically put literals at the end. 
```
let mut new_scalar_spec = scalars.drain(..).enumerate().map(|(old_col, e)| (e.is_literal(), old_col, e));
// this automatically puts literals at the end.
new_scalar_spec.sort();
(scalar_projection_to_reverse, new_scalars) = new_scalar_spec.into_iter().(|(_, old_col, e)| (old_col, e)).unzip();
let mut scalar_projection = vec![0; scalar_projection_to_reverse.len()];
for (new_col, old_col) in scalar_projection_to_reverse.into_iter().enumerate() {
   scalar_projection[old_col] = new_col;
}
let projection = (0..input_arity).collect::<Vec<usize>>().extend(scalar_projection.into_iter().map(|col| input_arity+ col));
```

Reply:
That code is missing the part that remaps the columns in the scalar expressions, which requires an imperative loop over `new_scalars`, which is just a re-ordered version of `scalars`. If we did everything in that imperative loop as this patch does, the final code is more concise IMO.

Reply:
I could avoid the cloning in the current version by using drain instead as follows:

```
diff --git a/src/transform/src/map_lifting.rs b/src/transform/src/map_lifting.rs
index 4afba49c2..5aad47135 100644
--- a/src/transform/src/map_lifting.rs
+++ b/src/transform/src/map_lifting.rs
@@ -281,14 +281,13 @@ impl LiteralLifting {
                         let mut new_scalars = Vec::new();
                         let mut projected_literals = Vec::new();
                         let mut projection = (0..input_arity).collect::<Vec<usize>>();
-                        for scalar in scalars.iter_mut() {
+                        for mut scalar in scalars.drain(..) {
                             if scalar.is_literal() {
                                 projection.push(first_literal_id + projected_literals.len());
-                                projected_literals.push(scalar.clone());
+                                projected_literals.push(scalar);
                             } else {
-                                let mut cloned_scalar = scalar.clone();
                                 // Propagate literals through expressions and remap columns.
-                                cloned_scalar.visit_mut(&mut |e| {
+                                scalar.visit_mut(&mut |e| {
                                     if let MirScalarExpr::Column(old_id) = e {
                                         let new_id = projection[*old_id];
                                         if new_id >= first_literal_id {
@@ -300,7 +299,7 @@ impl LiteralLifting {
                                     }
                                 });
                                 projection.push(input_arity + new_scalars.len());
-                                new_scalars.push(cloned_scalar);
+                                new_scalars.push(scalar);
                             }
                         }
                         new_scalars.extend(projected_literals);

```